internal/clonepool/clonepool.go

```diff
@@ -0,0 +1,245 @@
+package clonepool
+
+import (
+	"fmt"
+	"github.com/labstack/gommon/log"
+	"github.com/modfin/henry/slicez"
+	"github.com/modfin/zdap"
+	"github.com/modfin/zdap/internal"
+	"github.com/modfin/zdap/internal/cloning"
+	"github.com/modfin/zdap/internal/zfs"
+	"sync"
+	"time"
+)
+
+type ClonePool struct {
+	resource        internal.Resource
+	cloneContext    *cloning.CloneContext
+	ClonesAvailable int
+	claimLock       sync.Mutex
+}
+
+func NewClonePool(resource internal.Resource, cloneContext *cloning.CloneContext) ClonePool {
+	return ClonePool{resource: resource, cloneContext: cloneContext}
+}
+
+func (c *ClonePool) Start() {
+	go func() {
+		for {
+			time.Sleep(time.Second)
```

Comment:
Yes that's a good point. I'm not sure how much resources it will claim just doing the reads, but it'll prevent writes (like people trying to attach) a lot of the time. Especially if every resource has one of these jobs running. Another option could be to start keeping more information in memory, so `c.action` only reads from the dataset when any action is actually required. 

I think we still want to prune expired clones pretty often since we'll otherwise have a bunch of docker containers running, and I think those might really add up.

Reply:
if we add `gc chan struct{}` to the `ClonePool` struct, and change the `time.Sleep(time.Second)` into:
```
select {
case <-time.After(time.Hour):
case <-c.gc:
}
```

We can trig the `action` func on demand instead, and we can create a separate func like this, to make it simpler to trig this from other methods, like the `Claim` function:
```
func (c *ClonePool) trigGC() {
	select {
	case c.gc <- struct{}{}:
	case <-time.After(time.Millisecond):
	}
}
```

That way we will replenish and prune the pool as fast as they are claimed.

Reply:
`action` is still triggered hourly (or rather an hour after `action` were last run), so they may be left up to an hour after they have expired. But if someone are claiming or manually expiring clones, they will/can be pruned immediately.

Reply:
One option here could also be to have one global channel for all resources, with a single goroutine reading from it, so at most one action() would be running at one time. If we want to minimize the risk of them competing for resources. Not sure if it's needed though.

Reply:
Yes, but I still think it's a bad idea to have something like that to run every second or once per minute.
We could also keep track of the ttl when something is claimed, and if ttl < 10 minutes (or something like that) set another timer that can trig `action` after ttl have expired.

Reply:
Agreed

Reply:
As discussed offline let's do the channel solution (one channel per resource) but with a buffer of 1. Then we trigger this on every claim, every expire (based on a timer set on claim), and every new snap.